File: CODE/TEXT/JAVASCRIPT/processing.js
// -- IMPORTS

import md5 from 'md5';
import { parseDefText, throwParsingError } from './parsing.js';

// -- CONSTANTS

const
    isBrowser = ( typeof window !== 'undefined' && typeof window.document !== 'undefined' );

// -- FUNCTIONS

export function getNaturalTextComparison(
    firstText,
    secondText
    )
{
    let firstCharacterCount = firstText.length;
    let secondCharacterCount = secondText.length;
    let firstCharacterIndex = 0;
    let secondCharacterIndex = 0;

    while ( firstCharacterIndex < firstCharacterCount
            && secondCharacterIndex < secondCharacterCount )
    {
        let firstCharacter = firstText[ firstCharacterIndex ];
        let secondCharacter = secondText[ secondCharacterIndex ];

        if ( firstCharacter === secondCharacter )
        {
            ++firstCharacterIndex;
            ++secondCharacterIndex;
        }
        else
        {
            if ( firstCharacter >= '0'
                 && firstCharacter <= '9'
                 && secondCharacter >= '0'
                 && secondCharacter <= '9' )
            {
                let firstNumberText = '';
                let secondNumberText = '';

                while ( firstCharacterIndex < firstCharacterCount
                        && firstText[ firstCharacterIndex ] >= '0'
                        && firstText[ firstCharacterIndex ] <= '9' )
                {
                    firstNumberText += firstText[ firstCharacterIndex ];

                    ++firstCharacterIndex;
                }

                while ( secondCharacterIndex < secondCharacterCount
                        && secondText[ secondCharacterIndex ] >= '0'
                        && secondText[ secondCharacterIndex ] <= '9' )
                {
                    secondNumberText += secondText[ secondCharacterIndex ];

                    ++secondCharacterIndex;
                }

                return parseInt( firstNumberText ) - parseInt( secondNumberText );
            }
            else if ( firstCharacter < secondCharacter )
            {
                return -1;
            }
            else
            {
                return 1;
            }
        }
    }

    return firstCharacterCount - secondCharacterCount;
}

// ~~

export function getDefTextHash(
    text
    )
{
    return md5( text );
}

// ~~

export function getDefTextUuid(
    text
    )
{
    if ( text === '' )
    {
        return '';
    }
    else
    {
        let hash = md5( text );

        return (
            hash.slice( 0, 8 )
            + '-'
            + hash.slice( 8, 12 )
            + '-'
            + hash.slice( 12, 16 )
            + '-'
            + hash.slice( 16, 20 )
            + '-'
            + hash.slice( 20, 32 )
            );
    }
}

// ~~

export function getDefTextTuid(
    text
    )
{
    if ( text === '' )
    {
        return '';
    }
    else
    {
        let hash = md5( text );
        let tuid = '';

        if ( isBrowser )
        {
            let buffer = '';

            for ( let byteIndex = 0; byteIndex < hash.length; byteIndex += 2 )
            {
                buffer += String.fromCharCode( parseInt( hash.slice( byteIndex, byteIndex + 2 ), 16 ) );
            }
            
            tuid = btoa( buffer );
        }
        else
        {
            tuid = Buffer.from( hash, 'hex' ).toString( 'base64' );
        }
    
        return (
            tuid
                .replaceAll( '+', '-' )
                .replaceAll( '/', '_' )
                .replaceAll( '=', '' )
            );
    }
}

// ~~

export function processDefQuotedString(
    string,
    context,
    level
    )
{
    if ( string.startsWith( '#' ) )
    {
        return getDefTextUuid( string.slice( 1 ) );
    }
    else if ( string.startsWith( '%' ) )
    {
        return getDefTextTuid( string.slice( 1 ) );
    }
    else
    {
        return string;
    }
}

File: CODE/PYTHON/TEXT/processing.py
# -- IMPORTS

import hashlib;
import os;

# -- CONSTANTS

is_browser = False;

# -- FUNCTIONS

def get_natural_text_comparison(
    first_text,
    second_text
    ):

    first_character_count = len( first_text );
    second_character_count = len( second_text );
    first_character_index = 0;
    second_character_index = 0;

    while first_character_index < first_character_count and second_character_index < second_character_count:

        first_character = first_text[ first_character_index ];
        second_character = second_text[ second_character_index ];

        if first_character == second_character:

            first_character_index += 1;
            second_character_index += 1;

        else:

            if (
                first_character >= "0"
                and first_character <= "9"
                and second_character >= "0"
                and second_character <= "9"
                ):

                first_number_text = "";
                second_number_text = "";

                while (
                    first_character_index < first_character_count
                    and first_text[ first_character_index ] >= "0"
                    and first_text[ first_character_index ] <= "9"
                    ):

                    first_number_text += first_text[ first_character_index ];

                    first_character_index += 1;

                while (
                    second_character_index < second_character_count
                    and second_text[ second_character_index ] >= "0"
                    and second_text[ second_character_index ] <= "9"
                    ):

                    second_number_text += second_text[ second_character_index ];

                    second_character_index += 1;

                return int( first_number_text ) - int( second_number_text );

            elif first_character < second_character:

                return -1;

            else:

                return 1;

    return first_character_count - second_character_count;

# ~~

def get_def_text_hash(
    text
    ):

    return hashlib.md5( text.encode( "utf8" ) ).hexdigest();

# ~~

def get_def_text_uuid(
    text
    ):

    if text == "":

        return "";

    else:

        hash_value = hashlib.md5( text.encode( "utf8" ) ).hexdigest();

        return (
            hash_value[ 0:8 ]
            + "-"
            + hash_value[ 8:12 ]
            + "-"
            + hash_value[ 12:16 ]
            + "-"
            + hash_value[ 16:20 ]
            + "-"
            + hash_value[ 20:32 ]
        );

# ~~

def get_def_text_tuid(
    text
    ):

    if text == "":

        return "";

    else:

        hash_value = hashlib.md5( text.encode( "utf8" ) ).hexdigest();

        import base64;

        tuid = base64.b64encode( bytes.fromhex( hash_value ) ).decode( "utf8" );

        return tuid.replace( "+", "-" ).replace( "/", "_" ).replace( "=", "" );

# ~~

def process_def_quoted_string(
    string,
    context,
    level
    ):

    if string.startswith( "#" ):

        return get_def_text_uuid( string[ 1: ] );

    elif string.startswith( "%" ):

        return get_def_text_tuid( string[ 1: ] );

    else:

        return string;
